### DISCUSSION TITLE
zephyr compilation error when using the flash

### DISCUSSION THREAD

**Discussion by EduardoGoncalves1966** [zephyrproject-rtos/zephyr#78055] [General]
Hi
I am trying to build a program that writes to an area of flash and then reads it back to confirm.
However I get a compilation error:
"undefined reference to __device_dts_ord_107"

I have added to .conf file:
CONFIG_FLASH=y
CONFIG_FLASH_MAP=y

I have added to the .dts file:
chosen {
....
    zephyr,flash-controller = &flash0;
....
};

&flash0 {
    status = "okay";
    label = "FLASH_0";
}

The line of the code that seems to generate the problem is:
flash_dev = DEVICE_DT_GET(DT_NODELABEL(flash0));

Any ideas on how to solve this? Thanks.

**Comment by ndrs-pst** [comment]
@EduardoGoncalves1966 Based on the `undefined reference to __device_dts_ord_107` error, it is implied that node 107 is not available, which may be due to the driver file not being compiled.

Given `devicetree_generated.h`, `autoconf.h`, and `zephyr.dts`, we can gain more insight to solve the issue.

**Reply by EduardoGoncalves1966** [reply]
Hi
Thanks for the reply. I have attached the files you requested. I added ".txt" to the file name.

[autoconf.h.txt](https://github.com/user-attachments/files/16897240/autoconf.h.txt)
[devicetree_generated.h.txt](https://github.com/user-attachments/files/16897241/devicetree_generated.h.txt)
[zephyr.dts.txt](https://github.com/user-attachments/files/16897243/zephyr.dts.txt)

**Comment by ndrs-pst** [comment]
In short please refer to node 106 which is `flash` not `flash0`.

Also you can get a lot of insight in `flash_shell.c` too.

**Comment by EduardoGoncalves1966** [comment]
thanks. It compiled ok.

**Comment by EduardoGoncalves1966** [comment]
I have compiled the program that uses the flash but I am having other problems when running the program.
My program erases, writes and reads back from the flash using flash_erase(), flash_write() and flash_read() functions.

when calling flash_erase() it calls flash_stm32_valid_range() function in flash_stm32l5x.c file.

bool flash_stm32_valid_range(const struct device *dev, off_t offset,
			     uint32_t len,
			     bool write)
{
	if (stm32_flash_has_2_banks(dev) &&
			(CONFIG_FLASH_SIZE < STM32_SERIES_MAX_FLASH)) {
		/*
		 * In case of bank1/2 discontinuity, the range should not
		 * start before bank2 and end beyond bank1 at the same time.
		 * Locations beyond bank2 are caught by
		 * flash_stm32_range_exists.
		 */
bank2_offset = BANK2_OFFSET;
flash_size   = FLASH_SIZE  ;

		if ((offset < BANK2_OFFSET) &&
					(offset + len > FLASH_SIZE / 2)) {
			return 0;
		}
	}

	if (write && !flash_stm32_valid_write(offset, len)) {
		return false;
	}
	return flash_stm32_range_exists(dev, offset, len);
}

At this stage the defines

BANK2_OFFSET is 0x100000 (1 Mbyte)
FLASH_SIZE is 0x0

which is wrong, since we are using a STM32H533R microcontroller with 512Kbytes of Flash starting at =x08000000.

The .dts file defines the flash as:

&flash0 {
	partitions {
		compatible = "fixed-partitions";
		#address-cells = <1>;
		#size-cells = <1>;
 
		flash_all_partition: partition@8000000 {
			label = "flash_all";
			reg = <0x8000000 DT_SIZE_K(512)>;
		};
	};
};

The linker.ld has these definitions:

#define ROMABLE_REGION 				FLASH
#define RAMABLE_REGION 				RAM
#define ROM_ADDR 					0x8000000
#define ROM_SIZE					512K
#define RAM_ADDR					0x20000000
#define RAM_SIZE					272K
#define MPU_ALIGN(region_size)		. = ALIGN(4)
#define FLASH_SECTION_SIZE			1K

/*----------------------------------------------------------------*/
/*--------------------------- MEMORY -----------------------------*/
/*----------------------------------------------------------------*/
MEMORY
{
	FLASH (rx) : ORIGIN = ROM_ADDR, LENGTH = ROM_SIZE
	RAM   (wx) : ORIGIN = RAM_ADDR, LENGTH = RAM_SIZE
	LINKER_DT_REGIONS()
	IDT_LIST (wx) : ORIGIN = 0xFFFF7FFF, LENGTH = 32K
}

Dos anyone has some ideas what could be wrong?  Thanks.

**Comment by ndrs-pst** [comment]
According to this

https://github.com/zephyrproject-rtos/zephyr/blob/dc5f1bfb3f2eee3e3775175232d508660f5d6efe/arch/Kconfig#L233-L240

The chosen should be similar to this beforehand (zephyr.dts after build `samples/drivers/flash_shell` with nucleo_h533re`)

```
	chosen {
		zephyr,flash-controller = &flash;
		zephyr,entropy = &rng;
		zephyr,console = &usart2;
		zephyr,shell-uart = &usart2;
		zephyr,sram = &sram1;
		zephyr,flash = &flash0;
	};

**Reply by EduardoGoncalves1966** [reply]
I have this:

	chosen {
		 zephyr,console = &uart5;
		 zephyr,shell-uart = &usart1;
		 zephyr,sram = &sram1;
		 zephyr,flash-controller = &flash0;
		 elux_macs = &uart4;
	};

Does this suggests a problem?

**Reply by ndrs-pst** [reply]
Should be
```
zephyr,flash-controller = &flash;
zephyr,flash = &flash0;
```

Then you can re-check the value of CONFIG_FLASH_SIZE in `autoconf.h`
